clang/runtime/dpct-rt/include/dpl_extras/memory.h

```diff
@@ -670,6 +670,235 @@ template <typename T> class device_iterator : public device_pointer<T> {
+      oneapi::dpl::execution::is_execution_policy_v<decayed_policy_or_tag_t>;
+  static constexpr bool is_sys_tag_v =
+      ::std::is_base_of_v<sys_tag, std::decay_t<decayed_policy_or_tag_t>>;
+  static_assert(is_policy_v || is_sys_tag_v,
+                "Only oneDPL policies or system tags may be provided");
+
+  using type = ::std::conditional_t<is_policy_v, policy_conversion,
+                                    decayed_policy_or_tag_t>;
+};
+
+template <typename PolicyOrTag>
+using policy_or_tag_to_tag_t = typename policy_or_tag_to_tag<PolicyOrTag>::type;
+
+} // namespace internal
+
+// TODO: Improve this pointer wrapper and make an iterator adaptor
+// once we have the Boost library dependency
+template <typename Tag, typename T, typename Derived>
+class tagged_pointer_base {
+public:
+  using pointer = T *;
+  using difference_type = ::std::make_signed_t<std::size_t>;
+  tagged_pointer_base() = default;
+  tagged_pointer_base(T *ptr) : m_ptr(ptr) {}
+  operator const T *() const { return m_ptr; }
+  operator T *() { return m_ptr; }
+
+protected:
+  T *m_ptr;
+};
+
+// tagged_pointer provides a wrapper around a raw pointer type with a tag of the
+// location of the allocated memory. Standard pointer operations are supported
+// with this class.
+template <typename Tag, typename T = void>
+class tagged_pointer
+    : public tagged_pointer_base<Tag, T, tagged_pointer<Tag, T>> {
+  using super_t = tagged_pointer_base<Tag, T, tagged_pointer<Tag, T>>;
+
+public:
+  using value_type = T;
+  using difference_type = typename super_t::difference_type;
+  using pointer = typename super_t::pointer;
+  using reference = T &;
+  using iterator_category = std::random_access_iterator_tag;
+  using is_hetero = ::std::false_type;
+  using is_passed_directly = std::true_type;
+
+  tagged_pointer() : super_t() {}
+  tagged_pointer(T *ptr) : super_t(ptr) {}
+  T &operator[](difference_type idx) { return this->m_ptr[idx]; }
+  T &operator[](difference_type idx) const { return this->m_ptr[idx]; }
+  tagged_pointer operator+(difference_type forward) const {
+    return tagged_pointer{this->m_ptr + forward};
+  }
+  tagged_pointer operator-(difference_type backward) const {
+    return tagged_pointer{this->m_ptr - backward};
+  }
+  tagged_pointer operator++(int) {
+    tagged_pointer p(this->m_ptr);
+    ++this->m_ptr;
+    return p;
+  }
+  tagged_pointer operator--(int) {
+    tagged_pointer p(this->m_ptr);
+    --this->m_ptr;
+    return p;
+  }
+  tagged_pointer &operator++() {
+    ++this->m_ptr;
+    return *this;
+  }
+  tagged_pointer &operator--() {
+    --this->m_ptr;
+    return *this;
+  }
+  difference_type operator-(const tagged_pointer &it) const {
+    return this->m_ptr - it.m_ptr;
+  }
+  tagged_pointer &operator+=(difference_type forward) {
+    this->m_ptr = this->m_ptr + forward;
+    return *this;
+  }
+  tagged_pointer &operator-=(difference_type backward) {
+    this->m_ptr = this->m_ptr - backward;
+    return *this;
+  }
```

Comment:
`operator*` and `operator->` are also needed?

Reply:
I have added this, so we are not relying on conversion to pointer. I also updated the test to check `operator*` and `operator->`.

Reply:
Do we need some type conversion like:
```
  template<class OtherT>
  operator tagged_pointer<Tag, OtherT>();
```

Reply:
I don't think this operator is necessary since we already provide a raw pointer conversion to `otherT` which can then call the `tagged_pointer<Tag, OtherT>` constructor. This functionality is tested in `onedpl_test_tagged_pointer` in the test PR.

We also don't provide this operator in similar classes (ex: device_pointer).